File: android/app/src/main/kotlin/jp/co/bitz/tictactoe/scenes/Strings.kt
package jp.co.bitz.tictactoe.scenes

import android.content.Context
import jp.co.bitz.tictactoe.R
import jp.co.bitz.tictactoe.game.Mark

/**
 * Every user-facing string, looked up from `res/values/strings.xml` (`res/values-ja/strings.xml`
 * for Japanese) — see docs/GAME_DESIGN.md's "Localization" section for the full key table. No
 * scene ever embeds a string literal directly.
 */
public class Strings(
    private val context: Context,
) {
    public val menuTitle: String get() = context.getString(R.string.menu_title)
    public val difficultyEasy: String get() = context.getString(R.string.difficulty_easy)
    public val difficultyNormal: String get() = context.getString(R.string.difficulty_normal)
    public val difficultyHard: String get() = context.getString(R.string.difficulty_hard)

    public fun yourTurn(mark: Mark): String = context.getString(R.string.status_your_turn, mark.displayString)

    public val cpuThinking: String get() = context.getString(R.string.status_cpu_thinking)
    public val youWin: String get() = context.getString(R.string.status_you_win)
    public val cpuWins: String get() = context.getString(R.string.status_cpu_wins)
    public val draw: String get() = context.getString(R.string.status_draw)

    public fun scoreRow(
        x: Int,
        o: Int,
        draws: Int,
    ): String = context.getString(R.string.score_row, x, o, draws)

    public val newGame: String get() = context.getString(R.string.new_game)
}

/** "X"/"O" — never translated, see docs/GAME_DESIGN.md's "Localization" section. */
public val Mark.displayString: String get() = if (this == Mark.X) "X" else "O"

File: ios/TicTacToe/Scenes/Strings.swift
import Foundation

/// Every user-facing string, looked up from `Resources/Localizable.xcstrings` — see
/// `docs/GAME_DESIGN.md`'s "Localization" section for the full key table. No scene ever embeds a
/// string literal directly.
enum Strings {
    static var menuTitle: String {
        String(localized: "menu_title", defaultValue: "Tic-Tac-Toe")
    }

    static var difficultyEasy: String {
        String(localized: "difficulty_easy", defaultValue: "Easy")
    }

    static var difficultyNormal: String {
        String(localized: "difficulty_normal", defaultValue: "Normal")
    }

    static var difficultyHard: String {
        String(localized: "difficulty_hard", defaultValue: "Hard")
    }

    static func yourTurn(_ mark: Mark) -> String {
        let format = String(localized: "status_your_turn", defaultValue: "Your turn (%@)")
        return String(format: format, mark.displayString)
    }

    static var cpuThinking: String {
        String(localized: "status_cpu_thinking", defaultValue: "CPU thinking…")
    }

    static var youWin: String {
        String(localized: "status_you_win", defaultValue: "You win!")
    }

    static var cpuWins: String {
        String(localized: "status_cpu_wins", defaultValue: "CPU wins")
    }

    static var draw: String {
        String(localized: "status_draw", defaultValue: "Draw")
    }

    static func scoreRow(x: Int, o: Int, draws: Int) -> String {
        let format = String(localized: "score_row", defaultValue: "X: %d  O: %d  Draws: %d")
        return String(format: format, x, o, draws)
    }

    static var newGame: String {
        String(localized: "new_game", defaultValue: "New Game")
    }
}

extension Mark {
    /// `"X"`/`"O"` — never translated, see `docs/GAME_DESIGN.md`'s "Localization" section.
    var displayString: String { self == .x ? "X" : "O" }
}
